exonum/src/explorer/mod.rs

```diff
@@ -0,0 +1,976 @@
+/// # use exonum::helpers::Height;
+/// let blockchain = // ...
+/// #                sample_blockchain();
+/// let explorer = BlockchainExplorer::new(&blockchain);
+/// let block: BlockInfo = explorer.block(Height(1)).unwrap();
+/// assert_eq!(block.height(), Height(1));
+/// assert_eq!(block.len(), 3);
+///
+/// // Iterate over transactions in the block
+/// for tx in &block {
+///     println!("{:?}: {:?}", tx.location(), tx.content());
+/// }
+/// # } // main
+/// # #[cfg(not(feature = "doctests"))] fn main() {}
+/// ```
+///
+/// # JSON presentation
+///
+/// ```
+/// # #[macro_use] extern crate serde_json;
+/// # extern crate exonum;
+/// # #[cfg(feature = "doctests")] fn main() {
+/// # use exonum::helpers::Height;
+/// # use exonum::explorer::{BlockchainExplorer, BlockInfo};
+/// # use exonum::explorer::tests::sample_blockchain;
+/// # let blockchain = sample_blockchain();
+/// # let explorer = BlockchainExplorer::new(&blockchain);
+/// let block: BlockInfo = // ...
+/// #                      explorer.block(Height(1)).unwrap();
+/// assert_eq!(
+///     serde_json::to_value(&block).unwrap(),
+///     json!({
+///         // `Block` representation
+///         "block": block.block(),
+///         // Array of `Precommit`s
+///         "precommits": *block.precommits(),
+///         // Array of transaction hashes
+///         "txs": *block.transaction_hashes(),
+///     })
+/// );
+/// # } // main
+/// # #[cfg(not(feature = "doctests"))] fn main() {}
+/// ```
+#[derive(Debug)]
+pub struct BlockInfo<'a> {
+    explorer: &'a BlockchainExplorer<'a>,
+    block: Block,
+    precommits: RefCell<Option<Vec<Precommit>>>,
+    txs: RefCell<Option<Vec<Hash>>>,
+}
+
+impl<'a> BlockInfo<'a> {
+    fn new(explorer: &'a BlockchainExplorer, height: Height) -> Self {
+        let schema = Schema::new(&explorer.snapshot);
+        let block = {
+            let hashes = schema.block_hashes_by_height();
+            let blocks = schema.blocks();
+
+            let block_hash = hashes.get(height.0).expect(&format!(
+                "Block not found, height: {:?}",
+                height
+            ));
+            blocks.get(&block_hash).expect(&format!(
+                "Block not found, hash: {:?}",
+                block_hash
+            ))
+        };
+
+        BlockInfo {
+            explorer,
+            block,
+            precommits: RefCell::new(None),
+            txs: RefCell::new(None),
+        }
+    }
+
+    /// Returns the block header as recorded in the blockchain.
+    pub fn block(&self) -> &Block {
+        &self.block
+    }
+
+    /// Returns the number of transactions in this block.
+    pub fn len(&self) -> usize {
+        self.block.tx_count() as usize
+    }
+
+    /// Is this block empty (i.e., contains no transactions)?
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns a list of precommits for this block.
+    pub fn precommits(&self) -> Ref<[Precommit]> {
```

Comment:
Is this adequate complexity for the task? The reason I use it is to make `BlockInfo` load as little data from the blockchain from the get-go as possible in order to make it viable as an `Item` for `Blocks` iterator.